lifecycle.go

```diff
@@ -0,0 +1,105 @@
+package lampshade
+
+import (
+	"context"
+	"net"
+)
+
+// LifecycleListener allows lampshade users to listen to lampshade lifecycle events.
+type LifecycleListener interface {
+	OnSessionInit(context.Context) context.Context
```

Comment:
Yeah that was actually my original thought especially with `StreamLifecycleListener` -- let me play with that again. I got away from it because the holding state seemed to cause issues with traces and thread safety, but I think that's only the case if we store spans, not contexts.